PathFind.cpp

```diff
@@ -1,109 +1,58 @@
 // Copyright 2024 Aditya Behrani
+#include "include/algorithm.h"
+#include "include/pathfinder.h"
+#include "src/bfs.cpp"
+#include "src/dfs.cpp"
+#include "src/dijkstra.cpp"
+#include "src/graph.cpp"
 #include <fstream>
 #include <iostream>
+#include <memory>
 #include <queue>
+#include <string>
 #include <vector>
-#include "src/graph.cpp"
-#include "src/bfs.cpp"
-#include "src/dfs.cpp"
-#include "src/dijkstra.cpp"
-
-int START, GOAL;
 
-Graph* buildGraph(std::string filename) {
-    Graph *g = new Graph();
-
-    std::fstream infile(filename);
-    if (!infile) {
-        std::cerr << "Unable to open " << filename << "\n";
-        delete g;
-        return NULL;
-    }
-
-    int N;
-    infile >> N;
-    for (int i = 0; i < N; i++) {
-        g->addNode(i);
-    }
-
-     //  add edges to the list
-    int edges;
-    int u, v, w;
-    infile >> edges;
-
-    std::string line;
-    for (int i = 0; i < edges; i++) {
-        infile >> u >> v >> w;
-        g->addEdge(u, v, w);
-        g->addEdge(v, u, w);
+void printPath(std::vector<int> &path) {
+  if (path.size() == 0) {
+    std::cout << "No path exists\n";
+  } else if (path.size() > 0) {
+    for (int i = 0; i < path.size(); i++) {
+      std::cout << "Node " << i << ": " << path.at(i) << "\n";
     }
-
-    infile.close();
-    return g;
+  }
 }
 
-void promptInputs(
-    std::string &filename
-) {
-    std::cout << "What data file: ";
-    std::cin >> filename;
-
-    std::cout << "Starting Node: ";
-    std::cin >> START;
-
-    std::cout << "Goal Node: ";
-    std::cin >> GOAL;
-}
-
-void printPath(
-    std::vector<int>* path
-) {
-    if (path == NULL) {
-        std::cout << "No path exists\n";
-    } else if (path->size() > 0) {
-        for (int i = 0; i < path->size(); i++) {
-            std::cout << "Node " << i << ": " << path->at(i) << "\n";
-        }
-    }
-}
-
-int main() {
-    std::string filename;
-    promptInputs(filename);
-
-    Graph *g = buildGraph(filename);
-    // invalid file
-    if (g == NULL) {
-        return 1;
-    }
-
-    int selection;
-    std::cout << "Select your Algorithm:\n1: DFS\n2: BFS\n3: Dijkstra's" \
-              << "\nSelection: ";
-
-    std::cin >> selection;
-    std::cout << "\n";
-
-    std::vector<int>* path = NULL;
-
-    if (selection == 1) {
-        path = DFS::dfs(g, START, GOAL);
-    } else if (selection == 2) {
-        path = BFS::bfs(g, START, GOAL);
-    } else if (selection == 3) {
-        path = DIJKSTRA::dijkstra(g, START, GOAL);
-    } else {
-        std::cout << "Not a valid option\n";
-        delete path;
-        delete g;
-        return 0;
-    }
-
-    printPath(path);
-
-    // cleanup
-    delete path;
-    delete g;
-
-    return 0;
+int main(int argc, char *argv[]) {
+  std::string filename = argv[1];
+  int start = std::stoi(argv[2]), goal = std::stoi(argv[3]);
+  int selection = std::stoi(argv[4]);
+
+  Graph g = Graph();
+  int res = g.build(filename);
```

Comment:
Better than before, but still a strange pattern.
You're initializing a Graph, and then calling a `.build` method on it to actually build it.
You have two better options:
1. Build during initialization by passing the filename to the constructor instead of to a separate `.build` method, e.g. `Graph g = Graph(filename)`
2. Use the [Builder Design Pattern](https://stackoverflow.com/questions/29881135/difference-between-builder-pattern-and-constructor). Probably not necessary here, since you only have one argument (the filename to read).

Reply:
Went with option 1, also added some error handling for bad files.